### ISSUE TITLE
File store does not recover channels/messages on restart

### ISSUE THREAD

**Issue by tvalletta** [nats-io/nats-streaming-server#402]
I am new to Nats, but thanks for this great software.

I am running Nats Streaming Server in docker. I have the file store data mounted to my local file system. When I restart the container, the console shows that Nats file store has recovered 2 channels (I have published 3, see below), but my channels start at sequence 0 again and my durable subscriptions don't pick up any of the old messages. They start at 0 again.

Perhaps I have something misconfigured? Obviously it should pick up at the sequence that it left off before it restarted, right?

This is the server output:
```sh
[1] 2017/10/13 14:36:03.853845 [INF] STREAM: Starting nats-streaming-server[test-cluster] version 0.5.0
[1] 2017/10/13 14:36:03.853904 [INF] STREAM: ServerID: aMAiiAoPReCvI3KEO0Qgkw
[1] 2017/10/13 14:36:03.853909 [INF] STREAM: Go version: go1.7.6
[1] 2017/10/13 14:36:03.854409 [INF] Starting nats-server version 0.9.6
[1] 2017/10/13 14:36:03.854466 [INF] Starting http monitor on 0.0.0.0:8222
[1] 2017/10/13 14:36:03.854691 [INF] Listening for client connections on 0.0.0.0:4222
[1] 2017/10/13 14:36:03.854719 [INF] Server is ready
[1] 2017/10/13 14:36:03.890071 [INF] Recovering the state...
[1] 2017/10/13 14:36:03.902927 [INF] Recovered 2 channels
[1] 2017/10/13 14:36:04.156364 [INF] STREAM: Message store is FILE
[1] 2017/10/13 14:36:04.156411 [INF] STREAM: ---------- Store Limits ----------
[1] 2017/10/13 14:36:04.156416 [INF] STREAM: Channels:                  100 *
[1] 2017/10/13 14:36:04.156421 [INF] STREAM: --------- Channels Limits --------
[1] 2017/10/13 14:36:04.156426 [INF] STREAM:   Subscriptions:          1000 *
[1] 2017/10/13 14:36:04.156430 [INF] STREAM:   Messages     :       1000000 *
[1] 2017/10/13 14:36:04.156434 [INF] STREAM:   Bytes        :     976.56 MB *
[1] 2017/10/13 14:36:04.156438 [INF] STREAM:   Age          :     unlimited *
[1] 2017/10/13 14:36:04.156485 [INF] STREAM: ----------------------------------
```

I have published to three channels and have durable subscriptions on all three channels. When I hit `http://localhost:8222/streaming/channelsz?subs=1` I get the following output (before restart):

```json
{
  "cluster_id": "test-cluster",
  "server_id": "aMAiiAoPReCvI3KEO0Qgkw",
  "now": "2017-10-13T14:53:48.5888105Z",
  "offset": 0,
  "limit": 1024,
  "count": 5,
  "total": 5,
  "channels": [
    {
      "name": "hackweek",
      "msgs": 0,
      "bytes": 0,
      "first_seq": 0,
      "last_seq": 0
    },
    {
      "name": "hackweek/test/subject/one",
      "msgs": 2116,
      "bytes": 289765,
      "first_seq": 1,
      "last_seq": 2116,
      "subscriptions": [
        {
          "inbox": "_INBOX.6SK4TQE6VF8990MYP5D68V",
          "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0Qgtk",
          "durable_name": "test-durable-name",
          "is_durable": true,
          "max_inflight": 1,
          "ack_wait": 10,
          "last_sent": 2116,
          "pending_count": 1,
          "is_stalled": true
        },
        {
          "inbox": "_INBOX.6SK4TQE6VF8990MYP5D68V",
          "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0Qgtk",
          "durable_name": "test-durable-name",
          "is_durable": true,
          "max_inflight": 1,
          "ack_wait": 10,
          "last_sent": 2116,
          "pending_count": 1,
          "is_stalled": true
        }
      ]
    },
    {
      "name": "hackweek/test/subject/three",
      "msgs": 708,
      "bytes": 99701,
      "first_seq": 1,
      "last_seq": 708,
      "subscriptions": [
        {
          "inbox": "_INBOX.6SK4TQE6VF8990MYP5D6TZ",
          "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0Qgwg",
          "durable_name": "test-durable-name",
          "is_durable": true,
          "max_inflight": 1,
          "ack_wait": 10,
          "last_sent": 708,
          "pending_count": 0,
          "is_stalled": false
        },
        {
          "inbox": "_INBOX.6SK4TQE6VF8990MYP5D6TZ",
          "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0Qgwg",
          "durable_name": "test-durable-name",
          "is_durable": true,
          "max_inflight": 1,
          "ack_wait": 10,
          "last_sent": 708,
          "pending_count": 0,
          "is_stalled": false
        }
      ]
    },
    {
      "name": "hackweek/test/subject/two",
      "msgs": 1061,
      "bytes": 145230,
      "first_seq": 1,
      "last_seq": 1061,
      "subscriptions": [
        {
          "inbox": "_INBOX.6SK4TQE6VF8990MYP5D6JF",
          "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0QgvD",
          "durable_name": "test-durable-name",
          "is_durable": true,
          "max_inflight": 1,
          "ack_wait": 10,
          "last_sent": 1061,
          "pending_count": 0,
          "is_stalled": false
        },
        {
          "inbox": "_INBOX.6SK4TQE6VF8990MYP5D6JF",
          "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0QgvD",
          "durable_name": "test-durable-name",
          "is_durable": true,
          "max_inflight": 1,
          "ack_wait": 10,
          "last_sent": 1061,
          "pending_count": 0,
          "is_stalled": false
        }
      ]
    },
    {
      "name": "toElasticsearch",
      "msgs": 0,
      "bytes": 0,
      "first_seq": 0,
      "last_seq": 0
    }
  ]
}
```

When I hit `http://localhost:8222/streaming/clientsz?subs=1` I get the following output (before restart):

```json
{
  "cluster_id": "test-cluster",
  "server_id": "aMAiiAoPReCvI3KEO0Qgkw",
  "now": "2017-10-13T14:59:49.1928629Z",
  "offset": 0,
  "limit": 1024,
  "count": 2,
  "total": 2,
  "clients": [
    {
      "id": "test-ordered-subscriber-2",
      "hb_inbox": "_INBOX.6SK4TQE6VF8990MYP5D5YB",
      "subscriptions": {
        "hackweek/test/subject/one": [
          {
            "inbox": "_INBOX.6SK4TQE6VF8990MYP5D68V",
            "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0Qgtk",
            "durable_name": "test-durable-name",
            "is_durable": true,
            "max_inflight": 1,
            "ack_wait": 10,
            "last_sent": 2832,
            "pending_count": 0,
            "is_stalled": false
          }
        ],
        "hackweek/test/subject/three": [
          {
            "inbox": "_INBOX.6SK4TQE6VF8990MYP5D6TZ",
            "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0Qgwg",
            "durable_name": "test-durable-name",
            "is_durable": true,
            "max_inflight": 1,
            "ack_wait": 10,
            "last_sent": 948,
            "pending_count": 0,
            "is_stalled": false
          }
        ],
        "hackweek/test/subject/two": [
          {
            "inbox": "_INBOX.6SK4TQE6VF8990MYP5D6JF",
            "ack_inbox": "_INBOX.aMAiiAoPReCvI3KEO0QgvD",
            "durable_name": "test-durable-name",
            "is_durable": true,
            "max_inflight": 1,
            "ack_wait": 10,
            "last_sent": 1420,
            "pending_count": 0,
            "is_stalled": false
          }
        ]
      }
    },
    {
      "id": "test-publisher",
      "hb_inbox": "_INBOX.3EC0WC6U0DJD2AFZGJUQAI"
    }
  ]
}
```

When I hit `http://localhost:8222/streaming/storez` I get the following output (before restart):

```json
{
  "cluster_id": "test-cluster",
  "server_id": "jDeo9jCreQLdgbGjqSTwzx",
  "now": "2017-10-13T15:49:30.3746418Z",
  "type": "FILE",
  "limits": {
    "max_channels": 100,
    "max_msgs": 1000000,
    "max_bytes": 1024000000,
    "max_age": 0,
    "max_subscriptions": 1000
  },
  "total_msgs": 5621,
  "total_bytes": 773800
}
```

I am using this docker-compose configuration (version 0.5.0):

```yml
version: '3.1'

services:

  nats-streaming:
    image: nats-streaming
    restart: always
    container_name: nats-streaming
    ports:
      - '4222:4222'
      - '8222:8222'
    volumes:
      - './datastore:/datastore'
    command: -store file -dir datastore -m 8222
```

**Comment by kozlovic** [comment]
@tvalletta Thank you for your interest in NATS Streaming!

Unfortunately, the issue is caused by the naming of your channels. You have used some naming that logically maps to directories, but that is not supported by NATS Streaming (yet the server is not rejecting those channels).

The server's file store implementation uses a flat layout, all channels should be under the given data store directory (`-dir <datastore>`). So on recovery, the server simply recovers the channel `hackweek` which is empty. 

If you were to replace the `/` separator with the NATS token separator `.`, it should start to work the way you intended. Could you give it a try and report back?

**Comment by tylertreat** [comment]
@kozlovic Should we reject the creation of these channels if the names are invalid?

**Comment by kozlovic** [comment]
Yes, but `/` is invalid for FileStore implementation, but would not for memstore, or SQL store or possibly new store. So it means that using `/` would not be portable. I am wondering if the check should be in the server to reject for all type of stores.

**Comment by tylertreat** [comment]
Agreed, behavior should be consistent across stores.

**Comment by tvalletta** [comment]
You are indeed correct. When I change the channel names, it all works. ty

**Comment by kozlovic** [comment]
I have created issue #403 and corresponding fix #404 though. Thank you for the report!
